### ISSUE TITLE
CreateHostbuilder details

### ISSUE THREAD

**Issue by AndyPook** [marceln/net-generic-host-boilerplate#1]
I'm probably just missing something, so just a query...

This looks a lot like the builtin "worker" template (obvs yours have some more detail/structure).
My main query though is around `Program.CreateHostBuilder`. A lot of it seems to replicate what `CreateDefaultBuilder` does for you (logging, config. see https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L188 )

Yours is slightly different (ie appsettings being non-optional). But unless it calls `Clear` first, this will be adding the same files again, just with different options. It'll give the same end result, but the duplication might cause some confusion later (maybe).
The builtin also adds logging by default.

It would be "better" if it used `env.EnvironmentName` (like the builtin does) rather than going direct to env vars. It's more flexible.

---

As said at the top, I'm probably missing something, just bouncing some ideas. Would interested in your take

**Comment by marceln** [comment]
Hey Andy,

You're right on all points. It does replicate `CreateDefaultBuilder`. It does so because in real projects I plugin other frameworks in there. For example, I use Serilog instead of the standard logger. Also, besides that, I forget what the defaults are.  

I imagine this is useless for most, but when I was down to my 3rd or 4th console app in a span of few weeks I realized I need this to quickly start a new one. Might come in handy for some people that want to learn about the generic host. 

Config files are mandatory because I need them to be, especially the `Acceptance` and `Production` ones. These apps go into AWS Fargate containers via a CI/CD pipeline and I need it to not start if the config file corresponding to the target environment isn't there for some reason. 
About duplication, looking at `HostingHostBuilderExtensions.cs` you might be right. But I have no idea how to clear that source collection.

Thanks for the tip about `env.EnvironmentName`. I updated it. I used it like that so that anyone can quickly see what env var they need to define, but I guess the code is nicer this way.

**Comment by AndyPook** [comment]
`config.Sources.Clear()`

You just need to take care when fiddling with these things. Some are additive, some reset or add a whole extra set.
eg `AddLogging` (see https://github.com/dotnet/runtime/blob/c49c65df805304907185bbb4dab909672103ffef/src/libraries/Microsoft.Extensions.Logging/src/LoggingServiceCollectionExtensions.cs#L32 )
is also called via ConfigureLogging (also via the builtin). So calling it again via `ConfigureServices` can be problematic (maybe, depending 😬 ) So should (probably) be set up via an extra `ConfigureLogging`. This will allow you to fiddle with the original builder ie add to it, clear it ...

Anyway... just to say that the builtin bootstraping bits can trip you up if you don't play by its "expectations" and configure the builders it adds, rather than ending up adding extra ones (ie `AddLogging`)

A lot of how the Host Builder works is "non-obvious" (it's kind of a builder of builders) and it's super easy to dot into discovering the "wrong" mechanism.
Mostly things "work" cos the default container has a last added wins rule.

Finally, if you want to go down the rabbit hole... Config actually gets built twice. Once for the Host, once for the App (see https://github.com/dotnet/runtime/blob/c49c65df805304907185bbb4dab909672103ffef/src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs#L124). A few things are created from the host config, before it gets to setting up the app context.

Lots of fun 😁 
Anyway... mostly irrelevant (just geeking out)... what you've done will just-work(tm) 99.999% of the time but I've def been tripped up by some of this 😱

**Comment by ghelyar** [comment]
Yes I think most people should probably just use `dotnet new worker`

That said, if you want to use this then you might want to turn it into a proper project template that you can install and use with `dotnet new`. You could also base it on the worker template, which includes extra implicit usings etc.

**Comment by marceln** [comment]
Thanks a lot for the feedback, @AndyPook and @ghelyar. A project template sounds good, although I'm not sure how used that would be. I'll maybe give it a shot these days.

**Comment by billwheaton1715** [comment]
I’ll use it!  I don’t have a lot of boiler plate to go to and I need all the help I can get . You three are a godsend.  I’m always needing a starting point and ‘Console.writeLine(“hollow planet”);’ is a little lacking if you know what I mean.
